Create methodComponent::printComponentsMatching().

chrisdembia opened the pull request:
This PR introduces a method on Component to help users discover the full path names of components in their model. This came out of the discussion in #1025. To facilitate this method, I created a new ComponentList filter: ComponentFilterFullPathNameRegex.

@tkuchida @klshrinidhi please review!

tkuchida commented on OpenSim/Common/Component.h:
```diff
@@ -485,6 +485,15 @@ OpenSim_DECLARE_ABSTRACT_OBJECT(Component, Object);
+     * components on the right leg of a model (if all components on the right
+     * side have "_r" in their name).
+     * @returns The number of matches. */
+    int printComponentsMatching(const std::string& substring);
```
Should this be a const method?

klshrinidhi replied:
* Return type should be `unsigned`.
* Maybe be explicit that argument `substring` can be regex. If you don't want users to know, then maybe change implementation to not use regex because simple substring find is faster.
* Maybe add a code example to show how to call this thing and what result one might get.

chrisdembia replied:
Good points with (1) and (3). 

My intent with (2) is that advanced users can use ComponentFilterFullPathNameRegex directly, but that this method is a convenience when you just want to check for a substring (and is also available in python/java/matlab--std::regex does not get wrapped). I could write a new ComponentFilterFullPathNameContainsSubstring or add another method on Component that takes a regex. What do you think?

chrisdembia replied:
I have addressed your feedback:

1. Made the method const.
2. Changed return type to unsigned.
3. Added a code example with sample output.

tkuchida commented:
`</review>` Cool! :snowflake: Just the one comment. Looks like this will be really useful, especially once `findComponent()` has been made private.

klshrinidhi commented on OpenSim/Common/ComponentList.h:
```diff
@@ -76,6 +78,21 @@ class ComponentFilterMatchAll : public ComponentFilter {
+provided regular expression. */
+class ComponentFilterFullPathNameRegex : public ComponentFilter {
+public:
+    ComponentFilterFullPathNameRegex(const std::regex& expr) : _expr(expr) {}
```
Maybe accept a `string` rather than reference to `regex` itself. That way, user of this class need not create an object of type `regex` to use this class. It also avoids copying of `regex` object. They can just pass a string and internally, this class can either:
* Create a `regex` and keep it as private member.
* Create a use a `regex` on the fly having a `string` as private member.

chrisdembia replied:
How about if I create a second constructor that takes a string? I'd like to leave one constructor that takes a regex, as someone could conceivably want to [set their own flags](http://www.cplusplus.com/reference/regex/basic_regex/basic_regex/) on the regex (e.g., case-insensitive).

klshrinidhi replied:
I think keeping it simple will be better. May be accept only a string for now and always perform a case-insensitive search or have another parameter (with default) for case.

If complaints arise in future, update it accordingly.

chrisdembia replied:
Based on this comment and your other one, then perhaps I shouldn't use a regex at all? What do you think? I can make it just do a substring search. And in the future, we can add the regex variant if need be.

klshrinidhi replied:
If @tkuchida agrees, go ahead and remove regex. I don't know if it worth having or not. If you do want to have regex, using a simple string will be better.

chrisdembia replied:
I'll wait for @tkuchida to respond but I think I do prefer not using the regex... my regex is probably wrong! I didn't try to escape any characters that the user may have provided in their substring.

chrisdembia replied:
And I'm more inclined to have a weaker form of this feature than figure out how to get the regex correct.

tkuchida replied:
Good point about regex. I agree: simpler is best; we can add regex in the future if desired/necessary. (We may then also want to prohibit special characters in Component names as well.)

chrisdembia replied:
Great. Thanks for the cross reference.

chrisdembia replied:
I have addressed this feedback: I replaced this filter with one that just checks for a substring.

chrisdembia commented:
@tkuchida and @klshrinidhi I have addressed your feedback.

@tkuchida I also made `findComponent()` protected. Upon merging, please close #1025 if you think this PR addresses that issue.

Ready for review again.

tkuchida commented:
Looks great! :+1:
The only remaining issue that came up in #1025 was potentially needing/wanting a `hasComponent()` method.

chrisdembia commented:
> The only remaining issue that came up in #1025 was potentially needing/wanting a hasComponent() method.

I pushed a commit that adds this method.

tkuchida commented:
> I pushed a commit that adds this method.

Ok, great! LGTM.
#1025 can :boom: on merge.

chrisdembia commented:
Would someone like to merge this?

chrisdembia commented:
Thank you. 💥
